storage/objects/main.go

```diff
@@ -0,0 +1,196 @@
+		log.Fatal(err)
+	}
+	// create another bucket later to be used to copy objects from the prev one.
+	if err := createBucketIfNotexits(ctx, client, projectID, dstBucketName); err != nil {
+		log.Fatal(err)
+	}
+	name := fmt.Sprintf("gs://%v/%v", bucketName, objectName)
+
+	if err := write(client, bucketName, objectName); err != nil {
+		log.Fatalf("failed to write to the object (%v): %v", objectName, err)
+	}
+
+	data, err := read(client, bucketName, objectName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("contents of %v: %q\n", name, string(data))
+
+	attrs, err := attrs(client, bucketName, objectName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("object attributes of %v: %+v\n", name, attrs)
+
+	err = makePublic(client, bucketName, objectName)
+	if err != nil {
+		log.Fatalf("failed to make the object public: %v", err)
+	}
+
+	newName, err := move(client, bucketName, objectName)
+	if err != nil {
+		log.Fatalf("failed to move the object: %v", err)
+	}
+	objectName = newName
+
+	if err := delete(client, bucketName, objectName); err != nil {
+		log.Fatalf("failed to delete the object: %v", err)
+	}
+}
+
+func write(client *storage.Client, bucket, object string) error {
+	ctx := context.Background()
+	// [START upload_file]
+	f, err := os.Open("notes.txt")
+	if err != nil {
+		return err
+	}
+	wc := client.Bucket(bucket).Object(object).NewWriter(ctx)
+	if _, err = io.Copy(wc, f); err != nil {
+		return err
+	}
+	if err := f.Close(); err != nil {
+		return err
+	}
+	if err := wc.Close(); err != nil {
+		return err
+	}
+	// [END upload_file]
+	return nil
+}
+
+func read(client *storage.Client, bucket, object string) ([]byte, error) {
+	ctx := context.Background()
+	// [START download_file]
+	rc, err := client.Bucket(bucket).Object(object).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	data, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, err
+	}
+	if err := rc.Close(); err != nil {
+		return nil, err
+	}
+	return data, nil
+	// [END download_file]
+}
+
+func attrs(client *storage.Client, bucket, object string) (*storage.ObjectAttrs, error) {
+	ctx := context.Background()
+	// [START get_metadata]
+	o := client.Bucket(bucket).Object(object)
+	attrs, err := o.Attrs(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return attrs, nil
+	// [END get_metadata]
+}
+
+func makePublic(client *storage.Client, bucket, object string) error {
+	ctx := context.Background()
+	// [START public]
+	acl := client.Bucket(bucket).Object(object).ACL()
+	if err := acl.Set(ctx, storage.AllUsers, storage.RoleReader); err != nil {
+		return err
+	}
+	// [END public]
+	return nil
+}
+
+func move(client *storage.Client, bucket, object string) (string, error) {
+	ctx := context.Background()
+	// [START move_file]
+	dstName := object + "-rename"
+
+	src := client.Bucket(bucket).Object(object)
+	dst := client.Bucket(bucket).Object(dstName)
+
+	if _, err := src.CopyTo(ctx, dst, nil); err != nil {
+		return "", err
```

Comment:
What do you mean? It is your idea that the snippets need to be in their functions so can be testable.

This is generally a poor sample though. Because storage API doesn't support moving and you need to handle the case where deleting the original fails.